### PR TITLE
feat!: Type definition specification with env vars and JSON files

### PR DESCRIPTION
**Pull request by emostov**
closes: #387 

## changes

BREAKING CHANGE remove /config/types.json; Functionality replaced with SAS_SUBSTRATE_TYPES.

Additionally the following environment variables are added: 
- SAS_SUBSTRATE_TYPES_BUNDLE
- SAS_SUBSTRATE_TYPES_CHAIN
- SAS_SUBSTRATE_TYPES_SPEC

## motivation

This will allow users to update chain types when there is a delay for type updates in the apps-config package supply chain or if a chain's type definitions simply do not exist in the apps-config.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -142,20 +142,45 @@ For more information on our configuration manager visit its readme [here](https:
...
+Sidecar affords environment variables which allow the user to specify an absolute path to a JSON file
+that contains type definitions in the corresponding formats. Consult polkadot-js/api for more info on
+the type formats (see `RegisteredTypes`).
+
+**N.B** Types set from environment variables will override the corresponding types pulled from
+@polkadot/apps-config.
+
+-   `SAS_SUBSTRATE_TYPES_BUNDLE`: a bundle of types with versioning info, type aliases, derives, and
+    rpc definitions. Format: `OverrideBundleType` (see `typesBundle`).
+-   `SAS_SUBSTRATE_TYPES_CHAIN`: type definitions keyed by the `chainName`. Format: `Record<string, RegistryTypes>` (see `typesChain`).
```
Comment (dvdplm) [inline README.md:158]: Maybe these "(see `thing`)" could be links?

**Review by dvdplm — APPROVED** [review]
This seems to get the job done.

Do you think it'd be good long term to switch to using a proper config file, rather than using env variables? So we'd have `substrate-api-sidecar --config=my-config.[json|toml|yaml]` and have all the type overrides and other configuration as keys in there?

File: src/main.ts [hunk 2]
```diff
@@ -16,8 +17,9 @@
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 import { ApiPromise } from '@polkadot/api';
-import { typesBundle, typesChain, typesSpec } from '@polkadot/apps-config/api';
+import * as apps from '@polkadot/apps-config/api';
```
Comment (TarikGul) [inline src/main.ts:20]: Whats the reason for importing the whole path here? I definitely think what you did looks a lot cleaner and way more readable.

**Review by TarikGul — APPROVED** [review]
LGTM. I like how it gives the end user more flexibility.

**Comment by emostov** [comment]
> Do you think it'd be good long term to switch to using a proper config file, rather than using env variables? So we'd have substrate-api-sidecar --config=my-config.[json|toml|yaml] and have all the type overrides and other configuration as keys in there?

@dvdplm Right now the equivalent would be putting all your env vars in a `.env` file, like .`env.polkadot`. Then the user would start up sidecar with `NODE_ENV=polkadot yarn start`. Do you think a yml type config file would be an improvement to this?

**Comment by dvdplm** [comment]
> @dvdplm Right now the equivalent would be putting all your env vars in a `.env` file, like .`env.polkadot`. Then the user would start up sidecar with `NODE_ENV=polkadot yarn start`. Do you think a yml type config file would be an improvement to this?

I do have a distaste for ENV vars, but my opinion on this is definitely a minority. My gut reaction to `.env` files is "well, why not use a proper config file then?". I think env vars are opaque to users, fiddly to set, even fiddlier to unset, tricky to document properly, no namespaces so they grow really long and fugly. 
I think using a good config file format, like TOML (and to a lesser extent YAML) or even `.ini` has a psychological impact on developers to document and test their configs.

But yeah, ENV vars are really popular with many smart people so I tend to keep my opinions to myself. They're good from the devops POV for sure, and has good tooling on that end.

When I wrote my first comment what I had in mind was the ability to merge configs from several places: a decent set of defaults from substrate, refined further from polkadot and finally overrideable by users for their own chains. Such a hierarchy of config values is tractable using config files in a sane format; with ENV vars not so much?

**Comment by emostov** [comment]
> > @dvdplm Right now the equivalent would be putting all your env vars in a `.env` file, like .`env.polkadot`. Then the user would start up sidecar with `NODE_ENV=polkadot yarn start`. Do you think a yml type config file would be an improvement to this?
> 
> I do have a distaste for ENV vars, but my opinion on this is definitely a minority. My gut reaction to `.env` files is "well, why not use a proper config file then?". I think env vars are opaque to users, fiddly to set, even fiddlier to unset, tricky to document properly, no namespaces so they grow really long and fugly.
> I think using a good config file format, like TOML (and to a lesser extent YAML) or even `.ini` has a psychological impact on developers to document and test their configs.
> 
> But yeah, ENV vars are really popular with many smart people so I tend to keep my opinions to myself. They're good from the devops POV for sure, and has good tooling on that end.
> 
> When I wrote my first comment what I had in mind was the ability to merge configs from several places: a decent set of defaults from substrate, refined further from polkadot and finally overrideable by users for their own chains. Such a hierarchy of config values is tractable using config files in a sane format; with ENV vars not so much?

I really have minimal opinion on this, just what ever works for the most people. Happy to explore/implement different solutions in the future - I just would need a clear direction

File: src/main.ts (hunk 2 above)
Comment (emostov) [reply src/main.ts:20]: @TarikGul Just to clarify the difference between the keys with those names and the imports with those same names

**Pull request merged**